# Limpeza dos dados
## Roubo de Celulares

Com os dados obtidos  e já compilados, o próximo passo é realizar a limpeza do dataset
Isso consiste em:
 - Separar a cidade de São Bernardo do Campo
 - Verificar valores nulos
 - Campos duplicados
 - Remover informações que não serão úteis
 - Acertar ou remover informações erradas

### Importar as bibliotecas

As bibliotecas que serão utilizadas são:
 - `Pandas` - Manipulação do dataset
 - `Matplotlib` - Visualização dos dados
 - `Seaborn`- Visualização dos dados
 - `Numpy` - habilita propriedades matemáticas

In [1]:
#importando as bibliotecas
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import os


### Carregar o dataset de roubo de celulares

Carregar o dataset, verificar suas dimensões e ter as primeiras impressões

In [2]:
path = 'data_cleaned/'

directory = os.path.join(path) # Diretório onde estão salvos os arquivos .xls
i=0 # Marcador criado para definir se é o primeiro arquivo ou não
counter=0 #Contador de arquivos lidos


for root,dirs,files in os.walk(directory): # root - Caminho lido | dirs - diretórios existentes dentro do caminho lido | files - arquivos existentes no caminho lido
    
    #Percorrer todos os arquivos um a um, e ler somente os arquivos com extensão .xls
    for file in files:
        if file.endswith("2020.csv"):
            dftemp = pd.read_csv(path+file, low_memory=False) # leitura do arquivo .xls ("corrompido")
            print(path+file)
            if i==0: # Condição IF para determinar se é a primeira leitura e copiar o dataset para a variável "df"
                df = dftemp.copy()
                i = 1
                del(dftemp) # Deletar a variável para não sobrecarregar a memória
                
            else: # Para os demais arquivos realizar a junção com o primeiro arquivo
                df = df.append(dftemp, ignore_index=True)
                del(dftemp) # Deletar a variável para não sobrecarregar a memória
                
            linhas = df.shape[0]
            counter = counter + 1
            
print("Linhas totais: ", linhas)
print('\nTotal de arquivos concatenados:', counter)
print('\nDimensões do dataset: \n\t- {} Linhas\n\t- {} Colunas'.format(df.shape[0],df.shape[1]))

data_cleaned/furto_veiculo2010_2020.csv
data_cleaned/roubo_celular2010_2020.csv
data_cleaned/roubo_veiculo2010_2020.csv
Linhas totais:  6258784

Total de arquivos concatenados: 3

Dimensões do dataset: 
	- 6258784 Linhas
	- 57 Colunas


In [3]:
print('{} Linhas\n{} Colunas'.format(df.shape[0],df.shape[1]))
df.head() # Mostra as 5 primeiras linhas do dataset

6258784 Linhas
57 Colunas


ANO_BO  NUM_BO NUMERO_BOLETIM          BO_INICIADO           BO_EMITIDO  \
0    2009    4329      4329/2009  31/12/2009 23:58:08  01/01/2010 00:01:48   
1    2009    6595      6595/2009  31/12/2009 23:55:54  01/01/2010 00:02:38   
2    2010       1         1/2010  01/01/2010 00:05:13  01/01/2010 00:12:24   
3    2010       1         1/2010  01/01/2010 00:48:15  01/01/2010 00:51:39   
4    2010       1         1/2010  01/01/2010 00:58:51  01/01/2010 01:02:29   

  DATAOCORRENCIA HORAOCORRENCIA PERIDOOCORRENCIA DATACOMUNICACAO  \
0     31/12/2009          23:00          A NOITE      31/12/2009   
1     31/12/2009            NaN          A TARDE      31/12/2009   
2     31/12/2009          19:40          A NOITE      31/12/2009   
3     31/12/2009          23:50          A NOITE      01/01/2010   
4     31/12/2009          22:30          A NOITE      01/01/2010   

        DATAELABORACAO  ... DESCR_COR_VEICULO    DESCR_MARCA_VEICULO  \
0  31/12/2009 23:58:08  ...            Branco         VW/GOL PLUS MI   
1  31/12/2009 23:55:54  ...            Branco               VW/KOMBI   
2  01/01/2010 00:05:13  ...              Azul       FIAT/PALIO YOUNG   
3  01/01/2010 00:48:15  ...             Preta       HONDA/CG 125 FAN   
4  01/01/2010 00:58:51  ...            Branco  IMP/FIAT UNO MILLE SX   

  ANO_FABRICACAO ANO_MODELO DESCR_TIPO_VEICULO QUANT_CELULAR MARCA_CELULAR  \
0         1997.0     1997.0          AUTOMOVEL           NaN           NaN   
1         1974.0     1974.0          AUTOMOVEL           NaN           NaN   
2         2000.0     2001.0          AUTOMOVEL           NaN           NaN   
3         2006.0     2006.0          MOTOCICLO           NaN           NaN   
4         1997.0     1998.0          AUTOMOVEL           NaN           NaN   

  ROUBO_C ROUBO_V FURTO_V  
0       0       0       1  
1       0       0       1  
2       0       0       1  
3       0       0       1  
4       0       0       1  

[5 rows x 57 columns]

### Verificar dados nulos e ausentes

É possivel ver que pelo menos 18 colunas possuem mais de 80% de dados ausentes, a grande maioria relacionado a dados pessoais.

Essas as quais podem ser diretamente removidas do dataset.

Abaixo a lista ordenada em % dos primeiros 35 mais ausentes

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6258784 entries, 0 to 6258783
Data columns (total 57 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   ANO_BO                    int64  
 1   NUM_BO                    int64  
 2   NUMERO_BOLETIM            object 
 3   BO_INICIADO               object 
 4   BO_EMITIDO                object 
 5   DATAOCORRENCIA            object 
 6   HORAOCORRENCIA            object 
 7   PERIDOOCORRENCIA          object 
 8   DATACOMUNICACAO           object 
 9   DATAELABORACAO            object 
 10  BO_AUTORIA                object 
 11  FLAGRANTE                 object 
 12  NUMERO_BOLETIM_PRINCIPAL  object 
 13  LOGRADOURO                object 
 14  NUMERO                    object 
 15  BAIRRO                    object 
 16  CIDADE                    object 
 17  UF                        object 
 18  LATITUDE                  object 
 19  LONGITUDE                 object 
 20  DESCRICAOLOCAL          

Renomear a coluna `PERIDOOCORRENCIA` para `PERIODOOCORRENCIA`

In [5]:
df.rename(columns={'PERIDOOCORRENCIA':'PERIODOOCORRENCIA'}, inplace = True)

In [6]:
nulos = df.isnull().sum()*100
nulos = nulos / len(df)
nulos.sort_values(ascending=False).head(35)

PARENTESCO                  100.000000
RELACIONAMENTO               99.999105
PROFISSAO                    99.898383
GRAUINSTRUCAO                99.894484
NACIONALIDADE                99.857161
ESTADOCIVIL                  99.847526
NATURALIDADE                 99.838131
DATANASCIMENTO               99.835479
IDADE                        99.830334
SEXO                         99.790183
VITIMAFATAL                  99.786236
TIPOPESSOA                   99.786236
NATUREZAVINCULADA            99.786220
CORCUTIS                     99.786220
TIPOVINCULO                  99.786204
DESDOBRAMENTO                94.816932
NUMERO_BOLETIM_PRINCIPAL     93.829744
EXAME                        83.911731
QUANT_CELULAR                65.123193
MARCA_CELULAR                57.909604
DESCR_MARCA_VEICULO          49.574726
PLACA_VEICULO                49.427684
CIDADE_VEICULO               49.354108
DESCR_TIPO_VEICULO           49.289766
UF_VEICULO                   49.211668
DESCR_COR_VEICULO        

### Remover colunas e dados indesejados

Primeiramente avaliar entre as colunas que tem mais que 70% dos dados ausentes o que é possível eliminar e criar uma lista com esses campos

In [7]:
nulos.index[nulos > 0.7]

Index(['HORAOCORRENCIA', 'NUMERO_BOLETIM_PRINCIPAL', 'LOGRADOURO', 'BAIRRO',
       'LATITUDE', 'LONGITUDE', 'EXAME', 'DESDOBRAMENTO', 'STATUS',
       'TIPOPESSOA', 'VITIMAFATAL', 'NATURALIDADE', 'NACIONALIDADE', 'SEXO',
       'DATANASCIMENTO', 'IDADE', 'ESTADOCIVIL', 'PROFISSAO', 'GRAUINSTRUCAO',
       'CORCUTIS', 'NATUREZAVINCULADA', 'TIPOVINCULO', 'RELACIONAMENTO',
       'PARENTESCO', 'PLACA_VEICULO', 'UF_VEICULO', 'CIDADE_VEICULO',
       'DESCR_COR_VEICULO', 'DESCR_MARCA_VEICULO', 'ANO_FABRICACAO',
       'ANO_MODELO', 'DESCR_TIPO_VEICULO', 'QUANT_CELULAR', 'MARCA_CELULAR'],
      dtype='object')

Os campos pessoais podem ser diretamente eliminados, para os demais campos será analisado a real necessidade

In [8]:
eliminar = ['PARENTESCO', 'RELACIONAMENTO', 'NACIONALIDADE', 'PROFISSAO',
       'GRAUINSTRUCAO', 'NATURALIDADE', 'DATANASCIMENTO', 'ESTADOCIVIL',
       'IDADE', 'NATUREZAVINCULADA', 'TIPOPESSOA', 'SEXO', 'CORCUTIS',
       'TIPOVINCULO']

In [9]:
df.drop(eliminar, axis=1, inplace=True)

In [10]:
len(df.columns)

43

### Analisando demais colunas

Usar comando abaixo para ajustar configurações de visualização do pandas para explorar os dados das colunas

In [11]:
pd.set_option('display.max_columns', 60) # Define limite de 60 colunas na visualização
#pd.set_option('display.max_seq_items', 30) # Define limite de linhas na visualização

Selecionar somente os campos onde há vitima fatal e contar.

De acordo com o site da SSP-SP, a referência de quantidade de Boletins de Ocorrência deve ser pelo Número do Boletim (NUMERO_BOLETIM).


In [12]:
vitimafatal = df[df['VITIMAFATAL'].isnull()==False]
total_vf = len(vitimafatal['NUMERO_BOLETIM'].unique())
print(total_vf, 'vitimas fatais de 2010 à 2020')

2100 vitimas fatais de 2010 à 2020


Neste caso, houve 2100 vitimas fatais entre 2010 e 2020,

Por esta razão os dados serão mantidos, apenas a coluna `'VITIMAFATAL'` não será removida.

O campo `'EXAME'`, `'RUBRICA'`, `'ESPECIE'`, `'DESDOBRAMENTO'`, possuem informações que não são relevantes para a ánalise, porém os demais dados devem ser mantidos, removendos apenas as colunas.


In [13]:
print("EXAME")
print(df.EXAME.unique()[:10],'\n')
print("RUBRICA")
print(df.RUBRICA.unique()[:10],'\n')
print("ESPECIE")
print(df.ESPECIE.unique()[:10],'\n')
print("DESDOBRAMENTO")
print(df.DESDOBRAMENTO.unique()[:10])


EXAME
[nan 'IC' 'IC-IML' 'IML' 'SVO'] 

RUBRICA
['Furto (art. 155) - VEICULO'
 'Furto qualificado (art. 155, §4o.) - RESIDENCIA'
 'Furto qualificado (art. 155, §4o.) - VEICULO'
 'Fuga de local de acidente (Art. 305)'
 'Localização/Apreensão e Entrega de veículo' 'Choque'
 'Entrega de veículo localizado/apreendido'
 'Localização/Apreensão de veículo' 'Ato Infracional'
 'Localização/Apreensão de objeto'] 

ESPECIE
['Título II - Patrimônio (arts. 155 a 183)'
 'L 9503/97 - Código de Trânsito Brasileiro' 'Localização e/ou Devolução'
 'Acidente de trânsito' 'Ato infracional' 'Captura procurado' 'Pessoa'
 'Título XI - Administração pública (arts. 312 a 359-H)'
 'Título VIII - Incolumidade pública (arts. 250 a 285)'
 'Outros - não criminal'] 

DESDOBRAMENTO
[nan '§4o. Se o crime é cometido:'
 'Possuir, deter, portar, adquirir, fornecer, receber, ter em depósito...'
 'caput. Subtrair coisa móvel alheia, mediante grave ameaça ou violência a pessoa'
 'Nas mesmas penas incorrem quem (Par. único)'


In [14]:
eliminar = ["EXAME", "RUBRICA", "ESPECIE", "DESDOBRAMENTO"]

df.drop(eliminar, axis=1, inplace=True)

In [15]:
df.shape

(6258784, 39)

## Analise de variáveis

In [16]:
df.columns

Index(['ANO_BO', 'NUM_BO', 'NUMERO_BOLETIM', 'BO_INICIADO', 'BO_EMITIDO',
       'DATAOCORRENCIA', 'HORAOCORRENCIA', 'PERIODOOCORRENCIA',
       'DATACOMUNICACAO', 'DATAELABORACAO', 'BO_AUTORIA', 'FLAGRANTE',
       'NUMERO_BOLETIM_PRINCIPAL', 'LOGRADOURO', 'NUMERO', 'BAIRRO', 'CIDADE',
       'UF', 'LATITUDE', 'LONGITUDE', 'DESCRICAOLOCAL', 'SOLUCAO',
       'DELEGACIA_NOME', 'DELEGACIA_CIRCUNSCRICAO', 'STATUS', 'VITIMAFATAL',
       'PLACA_VEICULO', 'UF_VEICULO', 'CIDADE_VEICULO', 'DESCR_COR_VEICULO',
       'DESCR_MARCA_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO',
       'DESCR_TIPO_VEICULO', 'QUANT_CELULAR', 'MARCA_CELULAR', 'ROUBO_C',
       'ROUBO_V', 'FURTO_V'],
      dtype='object')

In [17]:
df.tail()

ANO_BO   NUM_BO NUMERO_BOLETIM          BO_INICIADO  \
6258779    2020  1441970   1441970/2020  30/09/2020 23:09:23   
6258780    2020  1441970   1441970/2020  30/09/2020 23:09:23   
6258781    2020  1440526   1440526/2020  30/09/2020 17:46:46   
6258782    2020  1439915   1439915/2020  30/09/2020 16:25:49   
6258783    2020     2624      2624/2020  30/09/2020 23:19:28   

                  BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
6258779  30/09/2020 23:07:52     30/09/2020          20:42           A NOITE   
6258780  30/09/2020 23:07:52     30/09/2020          20:42           A NOITE   
6258781  30/09/2020 23:08:45     30/09/2020          10:45        PELA MANHÃ   
6258782  30/09/2020 23:33:10     29/09/2020          11:20        PELA MANHÃ   
6258783  30/09/2020 23:37:33     30/09/2020          22:30           A NOITE   

        DATACOMUNICACAO       DATAELABORACAO    BO_AUTORIA FLAGRANTE  \
6258779      30/09/2020  30/09/2020 23:09:23  Desconhecida       Não   
6258780      30/09/2020  30/09/2020 23:09:23  Desconhecida       Não   
6258781      30/09/2020  30/09/2020 17:46:46  Desconhecida       Não   
6258782      30/09/2020  30/09/2020 16:25:49  Desconhecida       Não   
6258783      30/09/2020  30/09/2020 23:19:28  Desconhecida       Não   

        NUMERO_BOLETIM_PRINCIPAL                             LOGRADOURO  \
6258779                      NaN            RUA ANDRÉIA DA SILVA BARROS   
6258780                      NaN            RUA ANDRÉIA DA SILVA BARROS   
6258781                      NaN  AVENIDA RAIMUNDO PEREIRA DE MAGALHÃES   
6258782                      NaN           ESTRADA TURÍSTICA DO JARAGUÁ   
6258783                      NaN                         RODOVIA SP 332   

          NUMERO                          BAIRRO           CIDADE  UF  \
6258779    347.0  JARDIM TERRAS DE SANTO ANTONIO      HORTOLANDIA  SP   
6258780    347.0  JARDIM TERRAS DE SANTO ANTONIO      HORTOLANDIA  SP   
6258781  12495.0                   JARDIM MARILU          S.PAULO  SP   
6258782    700.0                    VILA JARAGUÁ          S.PAULO  SP   
6258783     45.0                  CRISTAL PARQUE  FRANCO DA ROCHA  SP   

                  LATITUDE          LONGITUDE  \
6258779  -22,9062783537479  -47,1787835739918   
6258780  -22,9062783537479  -47,1787835739918   
6258781        -23,4307817        -46,7203606   
6258782        -23,4880678        -46,7554815   
6258783        -23,2960414      -46,777100127   

                          DESCRICAOLOCAL                         SOLUCAO  \
6258779  Garagem ou abrigo de residência            BO PARA INVESTIGAÇÃO   
6258780  Garagem ou abrigo de residência            BO PARA INVESTIGAÇÃO   
6258781                      Via pública                BO PARA REGISTRO   
6258782                      Via pública  ENCAMINHAMENTO DP ÁREA DO FATO   
6258783                      Via pública  ENCAMINHAMENTO DP ÁREA DO FATO   

                   DELEGACIA_NOME DELEGACIA_CIRCUNSCRICAO     STATUS  \
6258779      DELEGACIA ELETRONICA    01º D.P. HORTOLÂNDIA  Consumado   
6258780      DELEGACIA ELETRONICA    01º D.P. HORTOLÂNDIA  Consumado   
6258781      DELEGACIA ELETRONICA          46º D.P. PERUS  Consumado   
6258782      DELEGACIA ELETRONICA          46º D.P. PERUS  Consumado   
6258783  DEL.POL.FRANCISCO MORATO    DEL.POL.FRANCO ROCHA  Consumado   

        VITIMAFATAL PLACA_VEICULO UF_VEICULO CIDADE_VEICULO DESCR_COR_VEICULO  \
6258779         NaN           NaN         SP            NaN             Preta   
6258780         NaN       DPL1580         SP       CAMPINAS          Vermelho   
6258781         NaN       EVI5936         SP       VALINHOS            Branco   
6258782         NaN       FSK6835         SP      SÃO PAULO            Branco   
6258783         NaN       EFL8943         SP        CAJAMAR          Vermelho   

            DESCR_MARCA_VEICULO  ANO_FABRICACAO  ANO_MODELO  \
6258779      HONDA/CG 125 TITAN             NaN         NaN   
6258780       YAMAHA/F

In [18]:
df.dtypes

ANO_BO                        int64
NUM_BO                        int64
NUMERO_BOLETIM               object
BO_INICIADO                  object
BO_EMITIDO                   object
DATAOCORRENCIA               object
HORAOCORRENCIA               object
PERIODOOCORRENCIA            object
DATACOMUNICACAO              object
DATAELABORACAO               object
BO_AUTORIA                   object
FLAGRANTE                    object
NUMERO_BOLETIM_PRINCIPAL     object
LOGRADOURO                   object
NUMERO                       object
BAIRRO                       object
CIDADE                       object
UF                           object
LATITUDE                     object
LONGITUDE                    object
DESCRICAOLOCAL               object
SOLUCAO                      object
DELEGACIA_NOME               object
DELEGACIA_CIRCUNSCRICAO      object
STATUS                       object
VITIMAFATAL                  object
PLACA_VEICULO                object
UF_VEICULO                  

As colunas `'DATACOMUNICACAO'`, `'DATAELABORACAO'` acabam sendo redundantes já que temos a data e horario do inicio e fim do BO, por este motivo as mesmas serão removidas

In [19]:
df.drop(['DATACOMUNICACAO', 'DATAELABORACAO'], axis=1, inplace=True)

### Colunas que devem ser convertidas para datas / inteiros / categoricos

In [20]:
datas = ['BO_INICIADO', 'BO_EMITIDO',
       'DATAOCORRENCIA', 'HORAOCORRENCIA']
inteiros = ['NUMERO', 'ANO_FABRICACAO', 'ANO_MODELO', 'QUANT_CELULAR']

categoricos = ['PERIODOOCORRENCIA', 'BO_AUTORIA', 'FLAGRANTE',
       'DESCRICAOLOCAL', 'SOLUCAO', 'STATUS',
       'VITIMAFATAL', 'DESCR_COR_VEICULO',
       'DESCR_TIPO_VEICULO']


#### Variáveis Categóricas

In [21]:
df[categoricos] = df[categoricos].astype("category")

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6258784 entries, 0 to 6258783
Data columns (total 37 columns):
 #   Column                    Dtype   
---  ------                    -----   
 0   ANO_BO                    int64   
 1   NUM_BO                    int64   
 2   NUMERO_BOLETIM            object  
 3   BO_INICIADO               object  
 4   BO_EMITIDO                object  
 5   DATAOCORRENCIA            object  
 6   HORAOCORRENCIA            object  
 7   PERIODOOCORRENCIA         category
 8   BO_AUTORIA                category
 9   FLAGRANTE                 category
 10  NUMERO_BOLETIM_PRINCIPAL  object  
 11  LOGRADOURO                object  
 12  NUMERO                    object  
 13  BAIRRO                    object  
 14  CIDADE                    object  
 15  UF                        object  
 16  LATITUDE                  object  
 17  LONGITUDE                 object  
 18  DESCRICAOLOCAL            category
 19  SOLUCAO                   category
 20  DE

#### Variáveis Númericas

In [23]:
df.select_dtypes(include=['integer', 'float']).columns

Index(['ANO_BO', 'NUM_BO', 'ANO_FABRICACAO', 'ANO_MODELO', 'ROUBO_C',
       'ROUBO_V', 'FURTO_V'],
      dtype='object')

Verificar se os valores de cada uma das colunas estão consistentes

##### ANO_BO

In [24]:
df['ANO_BO'].unique()

array([2009, 2010, 2011, 2012, 2017, 2013, 2016, 2014, 2015, 2018, 2019,
       2020], dtype=int64)

##### NUM_BO

In [25]:
df.loc[df['NUM_BO'] < 0]

Empty DataFrame
Columns: [ANO_BO, NUM_BO, NUMERO_BOLETIM, BO_INICIADO, BO_EMITIDO, DATAOCORRENCIA, HORAOCORRENCIA, PERIODOOCORRENCIA, BO_AUTORIA, FLAGRANTE, NUMERO_BOLETIM_PRINCIPAL, LOGRADOURO, NUMERO, BAIRRO, CIDADE, UF, LATITUDE, LONGITUDE, DESCRICAOLOCAL, SOLUCAO, DELEGACIA_NOME, DELEGACIA_CIRCUNSCRICAO, STATUS, VITIMAFATAL, PLACA_VEICULO, UF_VEICULO, CIDADE_VEICULO, DESCR_COR_VEICULO, DESCR_MARCA_VEICULO, ANO_FABRICACAO, ANO_MODELO, DESCR_TIPO_VEICULO, QUANT_CELULAR, MARCA_CELULAR, ROUBO_C, ROUBO_V, FURTO_V]
Index: []

##### NUMERO

In [26]:
df = df[(df['NUMERO'] != 'Título II - Patrimônio (arts. 155 a 183)') & (df['NUMERO'] != '96º D.P. MONÇÕES')]

In [27]:
for i in inteiros:
    print(i)
    df[i] = pd.to_numeric(df[i].replace(np.nan, 0), downcast='integer', errors='coerce')

NUMERO
ANO_FABRICACAO
ANO_MODELO
QUANT_CELULAR


Na coluna número é possivel ver que tem alguns números negativos, para isso vamos considerá-los como positivos

In [28]:
pd.set_option('display.float_format', lambda x: '%.0f' % x)

In [29]:
df['NUMERO'].describe()

count      6258782
mean         27959
std        5103881
min         -19739
25%              0
50%            100
75%            440
max     1998810785
Name: NUMERO, dtype: float64

In [30]:
len(df.loc[df['NUMERO'] < 0])

7

In [31]:
df['NUMERO'] = np.where((df['NUMERO'] < 0), abs(df['NUMERO']), df['NUMERO'])

In [32]:
df.loc[df['NUMERO'] < 0]

Empty DataFrame
Columns: [ANO_BO, NUM_BO, NUMERO_BOLETIM, BO_INICIADO, BO_EMITIDO, DATAOCORRENCIA, HORAOCORRENCIA, PERIODOOCORRENCIA, BO_AUTORIA, FLAGRANTE, NUMERO_BOLETIM_PRINCIPAL, LOGRADOURO, NUMERO, BAIRRO, CIDADE, UF, LATITUDE, LONGITUDE, DESCRICAOLOCAL, SOLUCAO, DELEGACIA_NOME, DELEGACIA_CIRCUNSCRICAO, STATUS, VITIMAFATAL, PLACA_VEICULO, UF_VEICULO, CIDADE_VEICULO, DESCR_COR_VEICULO, DESCR_MARCA_VEICULO, ANO_FABRICACAO, ANO_MODELO, DESCR_TIPO_VEICULO, QUANT_CELULAR, MARCA_CELULAR, ROUBO_C, ROUBO_V, FURTO_V]
Index: []

Também é possivel identificar diversos números muito maiores do que o encontrados normalmente em nossas ruas, mas fica impossivel saber qual seria a numeração correta

primeiramente será feito um filtro onde todos os números maiors do que 9 mil serão avaliados e contados

In [33]:
out_num = df['NUMERO'].loc[df['NUMERO'] > 9000].value_counts().sum()
out_num

77718

In [34]:
#20 primeiros registros com números maiores do que 9mil
df['NUMERO'].loc[df['NUMERO'] > 9000].value_counts().head(20)

99999    15463
9999      6957
10000     3818
12000     2192
15000     1743
11000     1496
11111     1252
13000     1004
22540      956
16000      932
14000      819
18000      668
20000      665
25000      653
17000      653
16741      645
10500      606
11500      555
9500       531
22000      469
Name: NUMERO, dtype: int64

In [35]:
out_num_p100 = out_num*100 / df.shape[0] 
print('{}% dos dados possuem número da rua maior do que 9 mil'.format(round(out_num_p100, 2)))

1.24% dos dados possuem número da rua maior do que 9 mil


Por ser uma quantia não tão significante, esses números considerados como número "1"


In [36]:
df['NUMERO'] = np.where((df['NUMERO'] > 9000), 1, df['NUMERO'])

Ao mesmo tempo existem mais de 1,73 milhão de registros com o número zero (27,64%), estes também serão passados para "1"

In [37]:
print(df['NUMERO'].loc[df['NUMERO'] == 0].value_counts().sum(), 'entradas com valor "zero"')
print(round(100*df['NUMERO'].loc[df['NUMERO'] == 0].value_counts().sum() / df.shape[0],2), '%')


1730148 entradas com valor "zero"
27.64 %


In [38]:
df['NUMERO'] = np.where((df['NUMERO'] == 0 ), 1, df['NUMERO'])

In [39]:
df['NUMERO'].describe()

count   6258782
mean        450
std         998
min           1
25%           1
50%         100
75%         402
max        9000
Name: NUMERO, dtype: float64

#####  ANO_FABRICACAO

Podemos verificar que há diversos anos de fabricação que não consistem com a realidade, uma vez que o primeiro automovel foi fabricado em 1886

Alguns deste valores como apenas "8" poderia representar "2008", outras como "95" poderiam ser "1995", mas como não tem como ter certeza, não podemos assumir isso logo de primeira.

O primeiro carro fabricado no Brasil foi no ano de 1956, pela Romi-sIetta


In [40]:
df['ANO_FABRICACAO'].unique()

array([1997, 1974, 2000, 2006,    0, 2008, 1994, 2007, 1989, 1985, 2002,
       2005, 1990, 1992, 1984, 1995, 1991, 1999, 2001, 2003, 1987, 2004,
       1996, 1988, 1983, 1993, 1982, 1998, 1975, 2009, 1979, 1976, 1986,
       1977, 1980, 1970, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1967, 1965, 2010,  200,  198, 1098, 2101, 1962, 2011, 1961,
         84, 1081, 1963, 1957,    7,    4, 1952, 1959,   89,  209, 2099,
       1960, 1841,   91, 2088, 1958,   98,   10,  978, 1951, 1899,  201,
       1928, 2012,  985, 2066, 2013,   12, 2998,   95, 8989, 2207,   97,
       2911, 2021, 2014, 3013, 2912, 1190,  202, 9394,   87,  992, 1085,
       1948,  211,  708, 1112, 1013, 2104, 2103, 1898, 2015, 2913, 1093,
       1191, 1195,  193, 1910,  213, 1193, 1079, 1883, 1083, 1944, 1889,
        910, 1111,  997, 1515, 1414,  304, 1194,  607, 1415, 1011, 1940,
       1199, 2206, 1010,   73, 1945, 1213, 1088,   99, 1914,    6,   96,
       1947, 7998,  809,   93, 1884, 1903, 2016,  2


- Primeiro passo será separar todas as entradas menor que 1950 e maior que 2020
- Depois contar qual a quantidade e a porcentagem destas entradas em relação aos demais
- Analisar em detalhes alguns destes valores por amostragem

In [41]:
#seleciona os campos onde Ano de Fabricação é menor que 1950 e maior que 2020
ano_fab = df.query('ANO_FABRICACAO < 1950 or ANO_FABRICACAO > 2020')

#Seleciona os campos onde Ano de Fabricação é diferente de zero
ano_fab = ano_fab.loc[df.ANO_FABRICACAO != 0]
print(len(ano_fab), 'Entradas onde Ano de fabricação é menor que 1950, maior que 2020 e diferente de zero')
print('{}% do total de entradas'.format(round(len(ano_fab)*100 / df.shape[0],2)))

467 Entradas onde Ano de fabricação é menor que 1950, maior que 2020 e diferente de zero
0.01% do total de entradas


In [42]:
# seleciona 20 amostras aleatórias para uma avaliação mais detalhada
ano_fab[['ANO_BO', 'PLACA_VEICULO', 'DESCR_MARCA_VEICULO', 'DESCR_TIPO_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO']].sample(n=20, random_state=111)

ANO_BO PLACA_VEICULO      DESCR_MARCA_VEICULO DESCR_TIPO_VEICULO  \
265403     2012       CFJ8507           H/HONDA CG 125          MOTOCICLO   
5881117    2017       PZN3268          FIAT/FIAT UNO S          AUTOMOVEL   
6232869    2020       LRF8342   I/BMW X5 M COMPETITION          AUTOMOVEL   
4655453    2011       MFE2419   SR/RODOTECNICA SRT TQ2       SEMI-REBOQUE   
1921541    2011       AMJ8941           M.BENZ/LS 1634    CAMINHÃO TRATOR   
1456835    2020       DYZ2042              H/HONDA 150          MOTOCICLO   
1994237    2012       AQL1447      VW/25.370 CLM T 6X2    CAMINHÃO TRATOR   
2946306    2015       FMQ1566                FIAT/FIAT          AUTOMOVEL   
2356315    2013       AWI7907       FIAT/MILLE ECONOMY          AUTOMOVEL   
2356317    2013       AWI7907       FIAT/MILLE ECONOMY          AUTOMOVEL   
5506106    2015       FMQ1566                FIAT/FIAT          AUTOMOVEL   
196073     2011       HLG1524  FIAT/PALIO FIRE ECONOMY          AUTOMOVEL   
4899998    2012       NZO3112                IMP/VOLVO           CAMINHÃO   
4640950    2011       GMQ6569            FORD/ESCORT L          AUTOMOVEL   
1234830    2018       IJM0929                 IMP/FORD          AUTOMOVEL   
123214     2010           HBT    HONDA/CBX 250 TWISTER          MOTOCICLO   
1971234    2012       NKA5421    REB/NOMA SR2E18RT1 CG       SEMI-REBOQUE   
4498844    2010       KJS7088     HONDA/XR 250 TORNADO          MOTOCICLO   
1009939    2016       FJT5760  CHEVROLET/CELTA 1.0L LT          AUTOMOVEL   
1296448    2018       BGS9643        IMP/VW GOL GL 1.8          AUTOMOVEL   

         ANO_FABRICACAO  ANO_MODELO  
265403              985        1985  
5881117            2107        2018  
6232869              13           0  
4655453             207        2007  
1921541            5004        2004  
1456835             200           0  
1994237             208        2008  
2946306             213         214  
2356315            7907        2012  
2356317            7907        2012  
5506106             213         214  
196073               10           0  
4899998            1011           0  
4640950              84          84  
1234830            1930        1930  
123214             1841           0  
1971234            2207        2008  
4498844             205           0  
1009939            1314        1314  
1296448              92          92

Conforme observado na tabela acima se analisado o campo `DESCR_MARCA_VEICULO`, `ANO_FABRICACAO` e `ANO_MODELO`, a maioria dos casos está claro que foram erros de digitação e poderiam ser facilmente arrumados.

No entanto por se tratar apenas de 0,01% de todos os registros esses valores serão considerados como "0"

In [43]:
#seleciona a coluna ANO_FABRICACAO onde o index é igual do dataset "ano_fab" e define os valores como "0"
df['ANO_FABRICACAO'].loc[ano_fab.index] = 0

c:\users\caior\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [44]:
df['ANO_FABRICACAO'].unique()

array([1997, 1974, 2000, 2006,    0, 2008, 1994, 2007, 1989, 1985, 2002,
       2005, 1990, 1992, 1984, 1995, 1991, 1999, 2001, 2003, 1987, 2004,
       1996, 1988, 1983, 1993, 1982, 1998, 1975, 2009, 1979, 1976, 1986,
       1977, 1980, 1970, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1967, 1965, 2010, 1962, 2011, 1961, 1963, 1957, 1952, 1959,
       1960, 1958, 1951, 2012, 2013, 2014, 2015, 2016, 1956, 2017, 1955,
       2018, 1954, 2019, 2020], dtype=int16)

#####  ANO_MODELO

Repetir os mesmos passos do `ANO_FABRICACAO` mas com `ANO_MODELO` até 2021

com o comando `df.ANO_MODELO.unique()` já é possivel identificar que há divergências nos valores 


In [45]:
df.ANO_MODELO.unique()

array([1997, 1974, 2001, 2006, 1998,    0, 2008, 2007, 1994, 1989, 1985,
       2002, 2005, 1990, 1992, 1984, 1996, 1991, 2000, 2009, 2003, 1987,
       1995, 2004, 1988, 1983, 1999, 1986, 1993, 1982, 1975, 1979, 1976,
       1977, 2010, 1980, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972,  125, 1600, 1970, 1967, 1965,  313, 5610, 2011, 1962, 4000,
       4013, 1218,  100, 1961,   84,   40, 1963, 1957,    8, 2020,    4,
       1952,  275, 1620, 1720, 1959, 1155,   89, 8140, 7630,  290, 1684,
       1634,  150, 1960, 2855,  500,    6,  250, 2012,   16, 5516,   91,
       3020,  580,  200,  285,  450, 1016, 1606, 1958,  400,   98, 1000,
       1951,  980, 1928, 9996,   10,    5, 2013, 2408,   12, 1111, 1518,
       2322,   97, 2207, 3514, 8150, 2208, 2066, 2014, 1780, 2913,   93,
       1190, 4283, 9394,  350, 1195,   88,   13, 3013,  815,    1, 9600,
       4600,  416,  292, 1948,  213,  112, 1013, 2015,  708,  750, 1112,
       1898,  265, 2103, 1093,  320, 1191, 2033, 19


- Primeiro passo será separar todas as entradas menor que 1950 e maior que 2021
- Depois contar qual a quantidade e a porcentagem destas entradas em relação aos demais
- Analisar em detalhes alguns destes valores por amostragem

In [46]:
#seleciona os campos onde Ano de Fabricação é menor que 1950 e maior que 2020
ano_mod = df.query('ANO_MODELO < 1950 or ANO_MODELO > 2021')

#Seleciona os campos onde Ano de Fabricação é diferente de zero
ano_mod = ano_mod.loc[df.ANO_MODELO != 0]
print(len(ano_mod), 'Entradas onde o ano do modelo é menor que 1950, maior que 2021 e diferente de zero')
print('{}% do total de entradas'.format(round(len(ano_mod)*100 / df.shape[0],2)))

1177 Entradas onde o ano do modelo é menor que 1950, maior que 2021 e diferente de zero
0.02% do total de entradas


In [47]:
# seleciona 20 amostras aleatórias para uma avaliação mais detalhada
ano_mod[['ANO_BO', 'PLACA_VEICULO', 'DESCR_MARCA_VEICULO', 'DESCR_TIPO_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO']].sample(n=20, random_state=111)

ANO_BO PLACA_VEICULO      DESCR_MARCA_VEICULO DESCR_TIPO_VEICULO  \
4755486    2011       MIJ9299                IMP/VOLVO           CAMINHÃO   
5480183    2015           NaN       I/SHINERAY XY 50 Q           MOTONETA   
2311407    2013       LLN9739  I/LR DISCOVERY 4 3.0 SE          AUTOMOVEL   
1667813    2010       BXC2856             REB/ANTONINI       SEMI-REBOQUE   
2013088    2012       HJR9607                VW/24.250           CAMINHÃO   
4970816    2013       NYR6395   SCANIA/P340 BAULER CAM    CAMINHÃO TRATOR   
4574541    2010       MFX5294        VOLVO/FH 440 6X2T       TRATOR RODAS   
3618602    2017           NaN          M.A./IMP TRATOR       TRATOR RODAS   
763734     2015           NaN     M.A./MASSEY-FERGUSON       TRATOR RODAS   
5527478    2015           NaN       I/SHINERAY XY 50 Q          CICLOMOTO   
2413752    2014       GSW1004               I/MERCEDES           CAMINHÃO   
551548     2013       GMZ1043         FORD/FORD F 4000           CAMINHÃO   
152542     2011       0000000          M.A./IMP TRATOR       TRATOR RODAS   
142674     2011       HMO8532                 GM/CELTA          AUTOMOVEL   
4793520    2012       GVJ6827  IMP/SCANIA T113H4X2 360    CAMINHÃO TRATOR   
1149768    2017           NaN         MA/MASSEY MF4265       TRATOR MISTO   
5239577    2014       FZX3635  CHEVROLET/ONIX 1.0MT LT          AUTOMOVEL   
326417     2012       COH1426            GM/KADETT GLS          AUTOMOVEL   
2658770    2014       EZL8012      I/KIA KM BETA BONGO           CAMINHÃO   
161548     2011       MSW2248              VW/CROSSFOX          AUTOMOVEL   

         ANO_FABRICACAO  ANO_MODELO  
4755486            2010         440  
5480183            2013        1014  
2311407            2011         211  
1667813            1993        1933  
2013088            2011        2425  
4970816               0         340  
4574541            2008         440  
3618602               0        6615  
763734             2000         275  
5527478            2015          50  
2413752            2011        1620  
551548             1989        4000  
152542             1980         285  
142674             2010        3020  
4793520               0          98  
1149768               0         265  
5239577            2013          10  
326417                0          97  
2658770               0          12  
161548             2009        1606

Como já "corrigimos" toda a coluna "ANO_FABRICACAO", vamos considerar que para todas essas entradas divergente que:
```python
ANO_MODELO = ANO_FABRICACAO
```


In [48]:
ano_mod['ANO_MODELO'] = ano_mod['ANO_FABRICACAO']

In [49]:
#seleciona a coluna ANO_FABRICACAO onde o index é igual do dataset "ano_fab" e define os valores como "0"
df['ANO_MODELO'].loc[ano_mod.index] = ano_mod['ANO_MODELO']

c:\users\caior\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [50]:
df.ANO_MODELO.unique()

array([1997, 1974, 2001, 2006, 1998,    0, 2008, 2007, 1994, 1989, 1985,
       2002, 2005, 1990, 1992, 1984, 1996, 1991, 2000, 2009, 2003, 1987,
       1995, 2004, 1988, 1983, 1999, 1986, 1993, 1982, 1975, 1979, 1976,
       1977, 2010, 1980, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1970, 1967, 1965, 2011, 1962, 1961, 1963, 1957, 2020, 1952,
       1959, 1960, 2012, 1958, 1951, 2013, 2014, 2015, 2016, 2017, 1956,
       2018, 1955, 2019, 1954, 2021], dtype=int16)

#####  QUANT_CELULAR

Como foi utilizado o comando 

```python
errors = 'coerce'
```
Os valores que estavam como inválidos passaram a ser considerados NaN, é possível verificar que apenas 7 entradas estavam com essa inconsistência, por este motivo as mesmas serão eliminadas



In [51]:
df.QUANT_CELULAR.isnull().sum()

7

In [52]:
df.dropna(subset = ['QUANT_CELULAR'], inplace=True)

Verificar as métricas estatisticas das quantidades de celulares roubados, apesar de aparecer alguns outliers, estes serão mantidos por poderem se tratar de roubo de carga

In [53]:
# resumo estatistico onde a quantidade de celular é diferente de zero
df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].describe()

count    2181781
mean           8
std         9972
min            1
25%            1
50%            1
75%            1
max     14728864
Name: QUANT_CELULAR, dtype: float64

Conforme tabela acima, é possivel ver que 75% das ocorrências acontecem com o roubo de apenas 1 celular,

Para verificação de uma amostragem, será considerado apenas quantidades que estão além dos 8 sigmas (99,9994%)

In [54]:
print('99,9994% das quantidades estão abaixo de:', df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].quantile(0.999994))

99,9994% das quantidades estão abaixo de: 5760.0


Apenas 12 entradas estão acima de 5760 unidades roubadas, e somente 2 entradas estão com valores discrepantes dos demais, portanto essas 2 entradas serão eliminadas

index:
 - 3411515
 - 3446741

In [55]:
df[['ANO_BO', 'NUM_BO', 'NUMERO_BOLETIM_PRINCIPAL', 'HORAOCORRENCIA', 'CIDADE', 'DESCRICAOLOCAL',
    'QUANT_CELULAR', 'MARCA_CELULAR']]\
.loc[df.QUANT_CELULAR > df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].quantile(0.999994)]

ANO_BO  NUM_BO NUMERO_BOLETIM_PRINCIPAL HORAOCORRENCIA  \
2496432    2014   12152                      NaN          16:00   
2508773    2014   14257                      NaN          15:00   
2620589    2014    4646                      NaN          22:20   
2724402    2014    4850                      NaN          00:30   
2898922    2015    1981                      NaN          06:15   
2900938    2015    2031        1981/2015 - 30127          06:15   
3285240    2016    1924                      NaN          18:30   
3411515    2017    1431                      NaN          01:30   
3446741    2017    1984                      NaN          10:30   
3742202    2018    2900                      NaN          16:55   
3743487    2018    1883       2900/2018 - 100908          16:55   
4337546    2020     594                      NaN          12:00   

                        CIDADE       DESCRICAOLOCAL  QUANT_CELULAR  \
2496432                S.PAULO          Via pública           9413   
2508773    S.BERNARDO DO CAMPO          Via pública           9675   
2620589                S.PAULO           Residência           9860   
2724402                DIADEMA          Via pública           9579   
2898922                DIADEMA          Via pública           9844   
2900938                DIADEMA          Via pública           9844   
3285240                S.PAULO  Comércio e serviços           9999   
3411515  FERRAZ DE VASCONCELOS          Via pública         168455   
3446741                 SUMARE          Via pública       14728864   
3742202             ARARAQUARA  Comércio e serviços           9999   
3743487             ARARAQUARA  Comércio e serviços           9999   
4337546            SERTAOZINHO  Comércio e serviços           9999   

              MARCA_CELULAR  
2496432             ALCATEL  
2508773             MOTORLA  
2620589            MOTOROLA  
2724402                 NaN  
2898922  SANSUNG GALAXY DUO  
2900938  SANSUNG GALAXY DUO  
3285240            DIVERSAS  
3411515             Samsung  
3446741                Asus  
3742202                 NaN  
3743487                 NaN  
4337546                 NaN

In [56]:
index_remove = df.loc[df['QUANT_CELULAR']> 100000].index
index_remove

Int64Index([3411515, 3446741], dtype='int64')

In [57]:
df.drop(index_remove, axis=0, inplace=True)

#### Datas

In [58]:
datas.append('PERIODOOCORRENCIA')

In [59]:
datas

['BO_INICIADO',
 'BO_EMITIDO',
 'DATAOCORRENCIA',
 'HORAOCORRENCIA',
 'PERIODOOCORRENCIA']

In [60]:
df[datas].sample(n=5)

BO_INICIADO           BO_EMITIDO DATAOCORRENCIA  \
4793239  20/11/2012 22:34:56  20/11/2012 23:05:11     20/11/2012   
3210727  17/04/2016 13:03:35  17/04/2016 13:03:35     15/04/2016   
2143119  13/09/2012 00:02:29  13/09/2012 02:20:54     12/09/2012   
4558766  01/07/2010 09:02:01  01/07/2010 09:08:26     01/07/2010   
449294   20/12/2013 23:36:33  20/12/2013 23:36:12     20/12/2013   

        HORAOCORRENCIA PERIODOOCORRENCIA  
4793239          19:30           A NOITE  
3210727          22:30           A NOITE  
2143119          22:30           A NOITE  
4558766          05:00      DE MADRUGADA  
449294           23:15   EM HORA INCERTA

In [61]:
df['BO_INICIADO'] = pd.to_datetime(df['BO_INICIADO'], format='%d/%m/%Y %H:%M:%S')
df['BO_EMITIDO'] = pd.to_datetime(df['BO_EMITIDO'], format='%d/%m/%Y %H:%M:%S')

In [62]:
df[datas].sample(n=5)

BO_INICIADO          BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA  \
360118  2012-06-15 14:39:39 2012-06-15 14:54:38     15/06/2012          10:20   
1380603 2019-03-21 08:03:12 2019-03-21 08:28:23     21/03/2019            NaN   
601620  2014-12-20 12:42:09 2014-12-20 12:59:42     20/12/2014          10:10   
4296349 2020-11-08 20:59:25 2020-11-08 21:16:38     23/10/2020          21:45   
5243868 2014-03-24 16:32:59 2014-03-24 16:49:01     24/03/2014          15:06   

        PERIODOOCORRENCIA  
360118         PELA MANHÃ  
1380603      DE MADRUGADA  
601620         PELA MANHÃ  
4296349           A NOITE  
5243868           A TARDE

In [63]:
df['PERIODOOCORRENCIA'].unique()

['A NOITE', 'A TARDE', 'DE MADRUGADA', 'PELA MANHÃ', 'EM HORA INCERTA']
Categories (5, object): ['A NOITE', 'A TARDE', 'DE MADRUGADA', 'PELA MANHÃ', 'EM HORA INCERTA']

In [64]:
df_datas = df.loc[df['PERIODOOCORRENCIA'] != 'EM HORA INCERTA']
df_datas.head()

ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO          BO_EMITIDO  \
0    2009    4329      4329/2009 2009-12-31 23:58:08 2010-01-01 00:01:48   
1    2009    6595      6595/2009 2009-12-31 23:55:54 2010-01-01 00:02:38   
2    2010       1         1/2010 2010-01-01 00:05:13 2010-01-01 00:12:24   
3    2010       1         1/2010 2010-01-01 00:48:15 2010-01-01 00:51:39   
4    2010       1         1/2010 2010-01-01 00:58:51 2010-01-01 01:02:29   

  DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA    BO_AUTORIA FLAGRANTE  \
0     31/12/2009          23:00           A NOITE  Desconhecida       Não   
1     31/12/2009            NaN           A TARDE  Desconhecida       Não   
2     31/12/2009          19:40           A NOITE  Desconhecida       Não   
3     31/12/2009          23:50           A NOITE  Desconhecida       Não   
4     31/12/2009          22:30           A NOITE  Desconhecida       Não   

  NUMERO_BOLETIM_PRINCIPAL                LOGRADOURO  NUMERO  \
0                      NaN       AV ESTRELA DA NOITE     643   
1                      NaN       RUA GERALDO MUZOLON     102   
2                      NaN  R MARIA GASTALDO CATELAN      60   
3                      NaN                       NaN       1   
4                      NaN          R COSTA DE LAVOS      75   

                 BAIRRO               CIDADE  UF LATITUDE LONGITUDE  \
0              JD. MAIA              S.PAULO  SP      NaN       NaN   
1  JD. PE. AUGUSTO SANI                  JAU  SP      NaN       NaN   
2            PQ. SELETA  S.BERNARDO DO CAMPO  SP      NaN       NaN   
3                   BNH            RIO CLARO  SP      NaN       NaN   
4        JD NOVA CARRAO              S.PAULO  SP      NaN       NaN   

  DESCRICAOLOCAL                         SOLUCAO  \
0    Via pública            BO PARA INVESTIGAÇÃO   
1    Via pública            BO PARA INVESTIGAÇÃO   
2    Via pública  ENCAMINHAMENTO DP ÁREA DO FATO   
3     Residência  APRECIAÇÃO DO DELEGADO TITULAR   
4    Via pública            BO PARA INVESTIGAÇÃO   

                 DELEGACIA_NOME       DELEGACIA_CIRCUNSCRICAO     STATUS  \
0      59º D.P. JARDIM DOS IPES      59º D.P. JARDIM DOS IPES  Consumado   
1           DEL.SEC.JAÚ PLANTÃO                  01º D.P. JAÚ  Consumado   
2  03º D.P. S.BERNARDO DO CAMPO  06º D.P. S.BERNARDO DO CAMPO  Consumado   
3     DEL.SEC.RIO CLARO PLANTÃO            03º D.P. RIO CLARO  Consumado   
4            41º D.P. VILA RICA            41º D.P. VILA RICA  Consumado   

  VITIMAFATAL PLACA_VEICULO UF_VEICULO        CIDADE_VEICULO  \
0         NaN       CIO9192         SP                SUZANO   
1         NaN       CQK4983         SP     CABRALIA PAULISTA   
2         NaN       DDW1129         SP  SAO BERNARDO DO CAMP   
3         NaN       KZW5136         SP             RIO CLARO   
4         NaN       CLS4059         SP             SAO PAULO   

  DESCR_COR_VEICULO    DESCR_MARCA_VEICULO  ANO_FABRICACAO  ANO_MODELO  \
0            Branco         VW/GOL PLUS MI            1997        1997   
1            Branco               VW/KOMBI            1974        1974   
2              Azul       FIAT/PALIO YOUNG            2000        2001   
3             Preta       HONDA/CG 125 FAN            2006        2006   
4            Branco  IMP/FIAT UNO MILLE SX            1997        1998   

  DESCR_TIPO_VEICULO  QUANT_CELULAR MARCA_CELULAR  ROUBO_C  ROUBO_V  FURTO_V  
0          AUTOMOVEL              0           NaN        0        0        1  
1          AUTOMOVEL              0           NaN        0        0        1  
2          AUTOMOVEL              0           NaN        0        0        1  
3          MOTOCICLO              0           NaN        0        0        1  
4          AUTOMOVEL              0           NaN        0        0        1

Verificar quantas entradas existem sem o registro da data de ocorrência, mas com horário conhecido

In [65]:
df_datas['DATAOCORRENCIA'].isnull().sum()

7

In [66]:
df_datas.loc[df_datas['DATAOCORRENCIA'].isnull()]

ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
4376360    2020    2050      2050/2020 2020-04-15 17:06:21   
4376361    2020    2050      2050/2020 2020-04-15 17:06:21   
6227941    2020     903       903/2020 2020-04-13 17:10:23   
6227945    2020    1236      1236/2020 2020-04-13 17:17:32   
6227946    2020    1236      1236/2020 2020-04-13 17:17:32   
6227947    2020    1236      1236/2020 2020-04-13 17:17:32   
6227948    2020    1236      1236/2020 2020-04-13 17:17:32   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
4376360 2020-04-15 17:42:54            NaN          21:00           A NOITE   
4376361 2020-04-15 17:42:54            NaN          21:00           A NOITE   
6227941 2020-04-13 18:05:12            NaN          15:50           A TARDE   
6227945 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227946 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227947 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227948 2020-04-13 18:24:58            NaN          12:16           A TARDE   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL  \
4376360  Desconhecida       Não                      NaN   
4376361  Desconhecida       Não                      NaN   
6227941  Desconhecida       Não                      NaN   
6227945  Desconhecida       Não                      NaN   
6227946  Desconhecida       Não                      NaN   
6227947  Desconhecida       Não                      NaN   
6227948  Desconhecida       Não                      NaN   

                          LOGRADOURO  NUMERO       BAIRRO  \
4376360          RUA DESIDERIO JORGE     195     VL NATAL   
4376361          RUA DESIDERIO JORGE     195     VL NATAL   
6227941           RUA 15 DE NOVEMBRO     800       CENTRO   
6227945  RUA valmir alves dos santos     202  VILA CENTRO   
6227946  RUA valmir alves dos santos     202  VILA CENTRO   
6227947  RUA valmir alves dos santos     202  VILA CENTRO   
6227948  RUA valmir alves dos santos     202  VILA CENTRO   

                        CIDADE  UF           LATITUDE          LONGITUDE  \
4376360        MOGI DAS CRUZES  SP  -23,5292832390769  -46,1823883556923   
4376361        MOGI DAS CRUZES  SP  -23,5292832390769  -46,1823883556923   
6227941           PIRASSUNUNGA  SP  -21,9957258847966  -47,4289556324068   
6227945  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227946  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227947  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227948  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   

        DESCRICAOLOCAL                         SOLUCAO  \
4376360    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
4376361    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227941    Via pública                BO PARA REGISTRO   
6227945    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227946    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227947    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227948    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   

                        DELEGACIA_NOME        DELEGACIA_CIRCUNSCRICAO  \
4376360       01º D.P. MOGI DAS CRUZES       01º D.P. MOGI DAS CRUZES   
4376361       01º D.P. MOGI DAS CRUZES       01º D.P. MOGI DAS CRUZES   
6227941          01º D.P. PIRASSUNUNGA          02º D.P. PIRASSUNUNGA   
6227945  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227946  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227947  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227948  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   

            STATUS VITIMAFATAL PLACA_VEICULO UF_VEICULO       CIDADE_VEICULO  \
4376360  Consumado         NaN           NaN        NaN                  NaN   
4376361  Consumado         NaN           NaN        NaN                  NaN   
6227941  C

In [67]:
ocorrencias_nulas = df_datas.loc[df_datas['DATAOCORRENCIA'].isnull()]
ocorrencias_nulas

ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
4376360    2020    2050      2050/2020 2020-04-15 17:06:21   
4376361    2020    2050      2050/2020 2020-04-15 17:06:21   
6227941    2020     903       903/2020 2020-04-13 17:10:23   
6227945    2020    1236      1236/2020 2020-04-13 17:17:32   
6227946    2020    1236      1236/2020 2020-04-13 17:17:32   
6227947    2020    1236      1236/2020 2020-04-13 17:17:32   
6227948    2020    1236      1236/2020 2020-04-13 17:17:32   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
4376360 2020-04-15 17:42:54            NaN          21:00           A NOITE   
4376361 2020-04-15 17:42:54            NaN          21:00           A NOITE   
6227941 2020-04-13 18:05:12            NaN          15:50           A TARDE   
6227945 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227946 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227947 2020-04-13 18:24:58            NaN          12:16           A TARDE   
6227948 2020-04-13 18:24:58            NaN          12:16           A TARDE   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL  \
4376360  Desconhecida       Não                      NaN   
4376361  Desconhecida       Não                      NaN   
6227941  Desconhecida       Não                      NaN   
6227945  Desconhecida       Não                      NaN   
6227946  Desconhecida       Não                      NaN   
6227947  Desconhecida       Não                      NaN   
6227948  Desconhecida       Não                      NaN   

                          LOGRADOURO  NUMERO       BAIRRO  \
4376360          RUA DESIDERIO JORGE     195     VL NATAL   
4376361          RUA DESIDERIO JORGE     195     VL NATAL   
6227941           RUA 15 DE NOVEMBRO     800       CENTRO   
6227945  RUA valmir alves dos santos     202  VILA CENTRO   
6227946  RUA valmir alves dos santos     202  VILA CENTRO   
6227947  RUA valmir alves dos santos     202  VILA CENTRO   
6227948  RUA valmir alves dos santos     202  VILA CENTRO   

                        CIDADE  UF           LATITUDE          LONGITUDE  \
4376360        MOGI DAS CRUZES  SP  -23,5292832390769  -46,1823883556923   
4376361        MOGI DAS CRUZES  SP  -23,5292832390769  -46,1823883556923   
6227941           PIRASSUNUNGA  SP  -21,9957258847966  -47,4289556324068   
6227945  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227946  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227947  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   
6227948  FERRAZ DE VASCONCELOS  SP      -23,543419231      -46,363991412   

        DESCRICAOLOCAL                         SOLUCAO  \
4376360    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
4376361    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227941    Via pública                BO PARA REGISTRO   
6227945    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227946    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227947    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   
6227948    Via pública  APRECIAÇÃO DO DELEGADO TITULAR   

                        DELEGACIA_NOME        DELEGACIA_CIRCUNSCRICAO  \
4376360       01º D.P. MOGI DAS CRUZES       01º D.P. MOGI DAS CRUZES   
4376361       01º D.P. MOGI DAS CRUZES       01º D.P. MOGI DAS CRUZES   
6227941          01º D.P. PIRASSUNUNGA          02º D.P. PIRASSUNUNGA   
6227945  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227946  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227947  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   
6227948  DEL.POL.FERRAZ DE VASCONCELOS  DEL.POL.FERRAZ DE VASCONCELOS   

            STATUS VITIMAFATAL PLACA_VEICULO UF_VEICULO       CIDADE_VEICULO  \
4376360  Consumado         NaN           NaN        NaN                  NaN   
4376361  Consumado         NaN           NaN        NaN                  NaN   
6227941  C

In [68]:
ocorrencias_nulas['BO_INICIADO'].dt.strftime("%d/%m/%Y")

4376360    15/04/2020
4376361    15/04/2020
6227941    13/04/2020
6227945    13/04/2020
6227946    13/04/2020
6227947    13/04/2020
6227948    13/04/2020
Name: BO_INICIADO, dtype: object

In [69]:
ocorrencias_nulas.index

Int64Index([4376360, 4376361, 6227941, 6227945, 6227946, 6227947, 6227948], dtype='int64')

In [70]:
df['DATAOCORRENCIA'][ocorrencias_nulas.index]

4376360    NaN
4376361    NaN
6227941    NaN
6227945    NaN
6227946    NaN
6227947    NaN
6227948    NaN
Name: DATAOCORRENCIA, dtype: object

In [71]:
df['DATAOCORRENCIA'][ocorrencias_nulas.index] = ocorrencias_nulas['BO_INICIADO'].dt.strftime("%d/%m/%Y")
df.iloc[ocorrencias_nulas.index]

<ipython-input-71-eb3cd89e72d6>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['DATAOCORRENCIA'][ocorrencias_nulas.index] = ocorrencias_nulas['BO_INICIADO'].dt.strftime("%d/%m/%Y")


ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
4376371    2020     580       580/2020 2020-04-15 17:18:14   
4376372    2020  476081    476081/2020 2020-04-15 18:04:00   
6227952    2020       9         9/2020 2020-04-13 18:30:49   
6227956    2020       9         9/2020 2020-04-13 18:30:49   
6227957    2020       9         9/2020 2020-04-13 18:30:49   
6227958    2020    1367      1367/2020 2020-04-13 16:04:31   
6227959    2020    1367      1367/2020 2020-04-13 16:04:31   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
4376371 2020-04-15 18:01:50     15/04/2020          10:00        PELA MANHÃ   
4376372 2020-04-15 18:02:54     15/04/2020          14:30           A TARDE   
6227952 2020-04-13 18:36:35     09/04/2020          10:00        PELA MANHÃ   
6227956 2020-04-13 18:36:35     09/04/2020          10:00        PELA MANHÃ   
6227957 2020-04-13 18:36:35     09/04/2020          10:00        PELA MANHÃ   
6227958 2020-04-13 18:59:37     13/04/2020          15:00           A TARDE   
6227959 2020-04-13 18:59:37     13/04/2020          15:00           A TARDE   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL  \
4376371  Desconhecida       Não                      NaN   
4376372  Desconhecida       Não                      NaN   
6227952  Desconhecida       Não     461624/2020 - 900020   
6227956  Desconhecida       Não     461624/2020 - 900020   
6227957  Desconhecida       Não     461624/2020 - 900020   
6227958     Conhecida       Sim                      NaN   
6227959     Conhecida       Sim                      NaN   

                                LOGRADOURO  NUMERO  \
4376371  RODOVIA PRESIDENTE DUTRA (BR 116)     133   
4376372             RUA PEDROSO DE CAMARGO      56   
6227952                RODOVIA FERNÃO DIAS     618   
6227956                RODOVIA FERNÃO DIAS     618   
6227957                RODOVIA FERNÃO DIAS     618   
6227958                  RUA DAS AVELEIRAS      45   
6227959                  RUA DAS AVELEIRAS      45   

                                   BAIRRO             CIDADE  UF  \
4376371                        AREA RURAL           CACAPAVA  SP   
4376372  CHÁCARA SANTO ANTÔNIO (ZONA SUL)            S.PAULO  SP   
6227952                         VILA MIRA  BRAGANCA PAULISTA  SP   
6227956                         VILA MIRA  BRAGANCA PAULISTA  SP   
6227957                         VILA MIRA  BRAGANCA PAULISTA  SP   
6227958                          ITAQUERA            S.PAULO  SP   
6227959                          ITAQUERA            S.PAULO  SP   

                  LATITUDE          LONGITUDE   DESCRICAOLOCAL  \
4376371  -23,1339389101186  -45,7493453491457      Via pública   
4376372        -23,6328118         -46,701995      Via Pública   
6227952        -23,1004084        -46,5692661  Rodovia/Estrada   
6227956        -23,1004084        -46,5692661  Rodovia/Estrada   
6227957        -23,1004084        -46,5692661  Rodovia/Estrada   
6227958  -23,5180565537814  -46,4534073906828      Via pública   
6227959  -23,5180565537814  -46,4534073906828      Via pública   

                                SOLUCAO             DELEGACIA_NOME  \
4376371  ENCAMINHAMENTO DP ÁREA DO FATO         08º D.P. GUARULHOS   
4376372                BO PARA REGISTRO       DELEGACIA ELETRONICA   
6227952                  BO PARA ADENDO  DEL.SEC.BRAGANÇA PAULISTA   
6227956                  BO PARA ADENDO  DEL.SEC.BRAGANÇA PAULISTA   
6227957                  BO PARA ADENDO  DEL.SEC.BRAGANÇA PAULISTA   
6227958               BO PARA FLAGRANTE   64º D.P. CID.AE CARVALHO   
6227959               BO PARA FLAGRANTE   64º D.P. CID.AE CARVALHO   

           DELEGACIA_CIRCUNSCRICAO     STATUS VITIMAFATAL PLACA_VEICULO  \
4376371           DEL.POL.CAÇAPAVA  Consumado         NaN       QPD8748   
4376372       11º D.P. SANTO AMARO  Consumado         NaN           NaN   
6227952  DEL.SEC.BRAGANÇA PAULISTA  Consumado         NaN       PRE9912   
6227956  DEL.SEC.BRAGANÇA PAUL

In [72]:
df.loc[df['DATAOCORRENCIA'].isnull()]

ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
4379449    2020    1267      1267/2020 2020-04-23 11:26:01   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
4379449 2020-04-23 13:09:47            NaN            NaN   EM HORA INCERTA   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL    LOGRADOURO  \
4379449  Desconhecida       Não                      NaN  RUA da barra   

         NUMERO       BAIRRO   CIDADE  UF           LATITUDE  \
4379449      52  PARELHEIROS  S.PAULO  SP  -23,7811257106842   

                 LONGITUDE DESCRICAOLOCAL               SOLUCAO  \
4379449  -46,7138911386842    Via pública  BO PARA INVESTIGAÇÃO   

               DELEGACIA_NOME DELEGACIA_CIRCUNSCRICAO     STATUS VITIMAFATAL  \
4379449  25º D.P. PARELHEIROS    25º D.P. PARELHEIROS  Consumado         NaN   

        PLACA_VEICULO UF_VEICULO CIDADE_VEICULO DESCR_COR_VEICULO  \
4379449           NaN        NaN            NaN               NaN   

        DESCR_MARCA_VEICULO  ANO_FABRICACAO  ANO_MODELO DESCR_TIPO_VEICULO  \
4379449                 NaN               0           0                NaN   

         QUANT_CELULAR MARCA_CELULAR  ROUBO_C  ROUBO_V  FURTO_V  
4379449              1      POSITIVO        1        0        0

Por ter apenas uma entrada sem data e hora do ocorrência, a mesma será eliminada

In [73]:
df.dropna(subset=['DATAOCORRENCIA'], axis=0, inplace=True)

In [74]:
df['HORAOCORRENCIA'].isnull().sum()

976481

In [75]:
df['HORAOCORRENCIA'] = np.where(df['HORAOCORRENCIA'].isnull(), df['BO_INICIADO'].dt.strftime('%H:%M'), df['HORAOCORRENCIA'])

In [76]:
df['HORAOCORRENCIA'].isnull().sum()

0

In [77]:
df['DATA_HORA_OCORRENCIA'] = pd.to_datetime(df['DATAOCORRENCIA'] + ' ' + df['HORAOCORRENCIA'], format='%d/%m/%Y %H:%M', errors='coerce')

In [78]:
df['DATA_HORA_OCORRENCIA'].isnull().sum()

76

In [79]:
erros_data = df.loc[df['DATA_HORA_OCORRENCIA'].isnull()]
erros_data

ANO_BO   NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
40446      2010      561       561/2010 2010-02-01 21:20:46   
51688      2010     1998      1998/2010 2010-03-08 01:54:35   
96374      2010     3748      3748/2010 2010-07-16 22:05:55   
1197626    2018  1268807   1268807/2018 2018-10-02 18:28:56   
1199414    2018  1287259   1287259/2018 2018-10-07 09:28:24   
...         ...      ...            ...                 ...   
6098055    2019   154558    154558/2019 2019-02-04 15:30:49   
6098173    2019   155807    155807/2019 2019-02-04 22:52:04   
6110900    2019   349376    349376/2019 2019-03-15 02:43:56   
6125040    2019   528591    528591/2019 2019-04-23 18:16:43   
6204103    2020  1973858   1973858/2020 2020-12-19 16:33:13   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
40446   2010-02-01 21:33:39     01/02/0210          03:30      DE MADRUGADA   
51688   2010-03-08 01:57:35     08/03/0210          00:30      DE MADRUGADA   
96374   2010-07-16 22:09:05     16/07/0201          22:05           A NOITE   
1197626 2018-10-02 18:28:57     02/01/0201          14:00           A TARDE   
1199414 2018-10-07 09:28:57     24/09/1018          22:00           A NOITE   
...                     ...            ...            ...               ...   
6098055 2019-02-04 15:31:19     23/01/1019          12:50           A TARDE   
6098173 2019-02-04 22:52:15     23/01/1019          06:35        PELA MANHÃ   
6110900 2019-03-15 02:44:39     02/03/1019          19:28           A NOITE   
6125040 2019-04-23 18:16:15     10/04/1019          21:25           A NOITE   
6204103 2020-12-19 16:33:16     17/12/0202          23:30           A NOITE   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL  \
40446    Desconhecida       Não                      NaN   
51688    Desconhecida       Não                      NaN   
96374    Desconhecida       Não                      NaN   
1197626  Desconhecida       Não                      NaN   
1199414  Desconhecida       Não                      NaN   
...               ...       ...                      ...   
6098055  Desconhecida       Não                      NaN   
6098173  Desconhecida       Não     151314/2019 - 900020   
6110900  Desconhecida       Não                      NaN   
6125040  Desconhecida       Não                      NaN   
6204103  Desconhecida       Não                      NaN   

                        LOGRADOURO  NUMERO           BAIRRO  \
40446                 AV SAPOPEMBA       1      VILA RENATA   
51688                  R GAL LECOR     100          VL ROSA   
96374            R DAS LARANJEIRAS     257           JARDIM   
1197626                 RUA RECIFE     130         GUAIAUNA   
1199414                 RUA JACOBE      54    TRANQUILIDADE   
...                            ...     ...              ...   
6098055  RUA DOUTOR AVELINO CHAVES     100  VILA LEOPOLDINA   
6098173         RUA DAS MARGARIDAS      80   VILA NOVA MAUÁ   
6110900               RUA ANHANGAI      44     CAMPO GRANDE   
6125040        RUA RAIMUNDO CORRÊA     178      VILA FEITAL   
6204103      RUA FRANCISCO PEIXOTO      79      CAMPO LIMPO   

                      CIDADE  UF           LATITUDE          LONGITUDE  \
40446                S.PAULO  SP                NaN                NaN   
51688    S.BERNARDO DO CAMPO  SP                NaN                NaN   
96374                S.ANDRE  SP                NaN                NaN   
1197626              S.PAULO  SP        -23,5295287        -46,5510861   
1199414            GUARULHOS  SP      -23,463946456  -46,5527904639999   
...                      ...  ..                ...                ...   
6098055              S.PAULO  SP  -23,5355159432473  -46,7279349189247   
6098173                 MAUA  SP  -23,6462428950176  -46,4521565589673   
6110900              S.PAULO  SP  -23,6884572369999    -46,68297746075   
6125040                 MAUA  SP        -23,6542931        -46,4206286   
6204103    

In [80]:
df.drop(erros_data.index, inplace = True)

É possivel identificar 4885 registros incoerentes em relação ao ano registrado da ocorrência com o ano do BO emitido, no caso como é impossivel saber a data correta estes registros serão eliminados

In [81]:
df.DATA_HORA_OCORRENCIA.dt.year.unique()

array([2009, 2010, 1910, 2008, 2006, 1997, 1976, 2001, 2007, 1987, 2004,
       1983, 1985, 1986, 2005, 1980, 1998, 1973, 1990, 2000, 1974, 2003,
       2002, 1965, 1977, 1970, 1999, 1991, 1982, 1971, 1996, 1995, 1940,
       1988, 2011, 1972, 1994, 1993, 1979, 1989, 1981, 1978, 1966, 1963,
       1948, 1954, 1955, 1992, 2012, 1956, 1960, 1968, 1912, 1975, 1953,
       1961, 1952, 1984, 1942, 2013, 1962, 1967, 1950, 1969, 1949, 1946,
       1913, 1959, 2014, 1939, 1964, 1957, 1914, 2015, 1915, 2016, 2017,
       1917, 2018, 1918, 2019, 1919, 2020, 1800, 1930, 1830, 1825, 1921,
       1958, 1937, 1916, 1945, 1947, 1951], dtype=int64)

In [82]:
#Selecionar todos os registros com ano da ocorrência inferior a 2009
dt_incoerente = df.loc[df.DATA_HORA_OCORRENCIA.dt.year < 2009]
dt_incoerente

ANO_BO   NUM_BO NUMERO_BOLETIM         BO_INICIADO  \
198        2010     1241      1241/2010 2010-01-02 15:04:21   
199        2010     1478      1478/2010 2010-01-02 18:14:42   
561        2010       12        12/2010 2010-01-04 14:28:12   
562        2010       12        12/2010 2010-01-04 14:28:12   
563        2010       12        12/2010 2010-01-04 14:28:12   
...         ...      ...            ...                 ...   
6238185    2020   759506    759506/2020 2020-06-10 04:07:11   
6246284    2020  1027458   1027458/2020 2020-07-25 06:11:38   
6250727    2020  1187402   1187402/2020 2020-08-19 23:51:34   
6250728    2020  1187402   1187402/2020 2020-08-19 23:51:34   
6258281    2020     2179      2179/2020 2020-09-28 10:25:26   

                 BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
198     2010-01-02 15:04:21     02/01/1910          15:04        PELA MANHÃ   
199     2010-01-02 15:04:21     02/01/1910          18:14        PELA MANHÃ   
561     2010-01-04 14:36:10     23/12/2008          14:28           A NOITE   
562     2010-01-04 14:36:10     23/12/2008          14:28           A NOITE   
563     2010-01-04 14:36:10     23/12/2008          14:28           A NOITE   
...                     ...            ...            ...               ...   
6238185 2020-06-10 04:09:57     09/06/1970          23:40           A NOITE   
6246284 2020-07-25 23:21:08     27/06/1977          21:39           A NOITE   
6250727 2020-08-19 23:51:41     19/08/1992          20:30           A NOITE   
6250728 2020-08-19 23:51:41     19/08/1992          20:30           A NOITE   
6258281 2020-09-28 10:48:37     15/07/2008          19:30           A NOITE   

           BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL  \
198      Desconhecida       Não                      NaN   
199      Desconhecida       Não                      NaN   
561      Desconhecida       Não      12463/2009 - 110429   
562      Desconhecida       Não      12463/2009 - 110429   
563      Desconhecida       Não      12463/2009 - 110429   
...               ...       ...                      ...   
6238185  Desconhecida       Não                      NaN   
6246284  Desconhecida       Não                      NaN   
6250727  Desconhecida       Não                      NaN   
6250728  Desconhecida       Não                      NaN   
6258281  Desconhecida       Não                      NaN   

                        LOGRADOURO  NUMERO            BAIRRO  \
198                   RUA AUSONIA,     504       VILA MAZZEI   
199                   RUA AUSONIA,     504       VILA MAZZEI   
561                R JOAO MESQUITA       1            CENTRO   
562                R JOAO MESQUITA       1            CENTRO   
563                R JOAO MESQUITA       1            CENTRO   
...                            ...     ...               ...   
6238185   ESTRADA VELHA DO JARAGUÁ     229      VILA JAGUARA   
6246284    RUA ANTONIO JOSE BENTES     454     CAPAO REDONDO   
6250727  RUA EPAMINONDAS RODRIGUES      20  PARQUE MARAJOARA   
6250728  RUA EPAMINONDAS RODRIGUES      20  PARQUE MARAJOARA   
6258281               RUA ARCTOTIS     271         SAO LUCAS   

                      CIDADE  UF           LATITUDE          LONGITUDE  \
198                SÃO PAULO  SP                NaN                NaN   
199                      NaN  SP                NaN                NaN   
561      S.JOSE DO RIO PRETO  SP                NaN                NaN   
562      S.JOSE DO RIO PRETO  SP                NaN                NaN   
563      S.JOSE DO RIO PRETO  SP                NaN                NaN   
...                      ...  ..                ...                ...   
6238185              S.PAULO  SP        -23,4227801        -46,7869952   
6246284              S.PAULO  SP  -23,6645684005454  -46,7919669462272   
6250727              S.ANDRE  SP  -23,6704261142121  -46,4963229792282   
6250728              S.ANDRE  SP  -23,6704261142121  -46,4963229792282   

In [83]:
df.drop(dt_incoerente.index, axis=0, inplace=True)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6253811 entries, 0 to 6258783
Data columns (total 38 columns):
 #   Column                    Dtype         
---  ------                    -----         
 0   ANO_BO                    int64         
 1   NUM_BO                    int64         
 2   NUMERO_BOLETIM            object        
 3   BO_INICIADO               datetime64[ns]
 4   BO_EMITIDO                datetime64[ns]
 5   DATAOCORRENCIA            object        
 6   HORAOCORRENCIA            object        
 7   PERIODOOCORRENCIA         category      
 8   BO_AUTORIA                category      
 9   FLAGRANTE                 category      
 10  NUMERO_BOLETIM_PRINCIPAL  object        
 11  LOGRADOURO                object        
 12  NUMERO                    int32         
 13  BAIRRO                    object        
 14  CIDADE                    object        
 15  UF                        object        
 16  LATITUDE                  object        
 17  LONGITUD

#### Reorganizar as colunas do dataset

In [85]:
df.columns.to_list()

['ANO_BO',
 'NUM_BO',
 'NUMERO_BOLETIM',
 'BO_INICIADO',
 'BO_EMITIDO',
 'DATAOCORRENCIA',
 'HORAOCORRENCIA',
 'PERIODOOCORRENCIA',
 'BO_AUTORIA',
 'FLAGRANTE',
 'NUMERO_BOLETIM_PRINCIPAL',
 'LOGRADOURO',
 'NUMERO',
 'BAIRRO',
 'CIDADE',
 'UF',
 'LATITUDE',
 'LONGITUDE',
 'DESCRICAOLOCAL',
 'SOLUCAO',
 'DELEGACIA_NOME',
 'DELEGACIA_CIRCUNSCRICAO',
 'STATUS',
 'VITIMAFATAL',
 'PLACA_VEICULO',
 'UF_VEICULO',
 'CIDADE_VEICULO',
 'DESCR_COR_VEICULO',
 'DESCR_MARCA_VEICULO',
 'ANO_FABRICACAO',
 'ANO_MODELO',
 'DESCR_TIPO_VEICULO',
 'QUANT_CELULAR',
 'MARCA_CELULAR',
 'ROUBO_C',
 'ROUBO_V',
 'FURTO_V',
 'DATA_HORA_OCORRENCIA']

In [86]:
df = df[['ANO_BO',
 'NUM_BO',
 'NUMERO_BOLETIM',
 'BO_INICIADO',
 'BO_EMITIDO',
 'DATA_HORA_OCORRENCIA',
 'DATAOCORRENCIA',
 'HORAOCORRENCIA',
 'PERIODOOCORRENCIA',
 'BO_AUTORIA',
 'FLAGRANTE',
 'NUMERO_BOLETIM_PRINCIPAL',
 'LOGRADOURO',
 'NUMERO',
 'BAIRRO',
 'CIDADE',
 'UF',
 'LATITUDE',
 'LONGITUDE',
 'DESCRICAOLOCAL',
 'SOLUCAO',
 'DELEGACIA_NOME',
 'DELEGACIA_CIRCUNSCRICAO',
 'STATUS',
 'VITIMAFATAL',
 'PLACA_VEICULO',
 'UF_VEICULO',
 'CIDADE_VEICULO',
 'DESCR_COR_VEICULO',
 'DESCR_MARCA_VEICULO',
 'ANO_FABRICACAO',
 'ANO_MODELO',
 'DESCR_TIPO_VEICULO',
 'QUANT_CELULAR',
 'MARCA_CELULAR',
 'ROUBO_C',
 'ROUBO_V',
 'FURTO_V']].copy()

In [87]:
df.sample()

ANO_BO  NUM_BO NUMERO_BOLETIM         BO_INICIADO          BO_EMITIDO  \
931128    2016    3416      3416/2016 2016-12-31 19:05:12 2016-12-31 19:12:32   

       DATA_HORA_OCORRENCIA DATAOCORRENCIA HORAOCORRENCIA PERIODOOCORRENCIA  \
931128  2016-12-31 19:05:00     31/12/2016          19:05      DE MADRUGADA   

          BO_AUTORIA FLAGRANTE NUMERO_BOLETIM_PRINCIPAL LOGRADOURO  NUMERO  \
931128  Desconhecida       Não                      NaN     RUA UM    1219   

                   BAIRRO  CIDADE  UF           LATITUDE          LONGITUDE  \
931128  MORRO SANTA MARIA  SANTOS  SP  -23,9381108953088  -46,3566924452007   

       DESCRICAOLOCAL                         SOLUCAO  DELEGACIA_NOME  \
931128    Via pública  ENCAMINHAMENTO DP ÁREA DO FATO  DEL.SEC.SANTOS   

       DELEGACIA_CIRCUNSCRICAO     STATUS VITIMAFATAL PLACA_VEICULO  \
931128         05º D.P. SANTOS  Consumado         NaN       BIM7375   

       UF_VEICULO CIDADE_VEICULO DESCR_COR_VEICULO   DESCR_MARCA_VEICULO  \
931128         SP         SANTOS             Preta  IMP/FIAT UNO CSL 1.6   

        ANO_FABRICACAO  ANO_MODELO DESCR_TIPO_VEICULO  QUANT_CELULAR  \
931128            1992        1992          AUTOMOVEL              0   

       MARCA_CELULAR  ROUBO_C  ROUBO_V  FURTO_V  
931128           NaN        0        0        1

In [88]:
df.shape

(6253811, 38)

#### Periodo Ocorrência 

É possivel perceber que a coluna `PERIODOOCORRENCIA` não condiz com o horário registrado

In [89]:
dfmad = df.loc[df.PERIODOOCORRENCIA == 'DE MADRUGADA']
dfman = df.loc[df.PERIODOOCORRENCIA == 'PELA MANHÃ']
dftar = df.loc[df.PERIODOOCORRENCIA == 'A TARDE']
dfnoi = df.loc[df.PERIODOOCORRENCIA == 'A NOITE']
dfinc = df.loc[df.PERIODOOCORRENCIA == 'EM HORA INCERTA']

df.PERIODOOCORRENCIA.value_counts()

A NOITE            2499783
PELA MANHÃ         1412263
A TARDE            1252392
DE MADRUGADA        936149
EM HORA INCERTA     153224
Name: PERIODOOCORRENCIA, dtype: int64

In [90]:
print('Madrugada: ', dfmad.DATA_HORA_OCORRENCIA.dt.hour.unique())
print('Manhã: ', dfman.DATA_HORA_OCORRENCIA.dt.hour.unique())
print('Tarde: ', dftar.DATA_HORA_OCORRENCIA.dt.hour.unique())
print('Noite: ', dfnoi.DATA_HORA_OCORRENCIA.dt.hour.unique())
print('Incerta: ', dfinc.DATA_HORA_OCORRENCIA.dt.hour.unique())

Madrugada:  [ 0  1  2  3  4  7  5 10 12 13 14 15 16 18  9 21 23  6 19 22  8 11 20 17]
Manhã:  [ 8  7  9 11 15 16 17  6 10 12 13 20 22 14 18 21  0 19  1 23  3  2  4  5]
Tarde:  [23 14 12 15 17 20 21 22 13  6 11 16 19  0 18 10  1  2  9  8  3  7  5  4]
Noite:  [23 19 22  1  2  8 18 20 21  0  6 10 16 11 15  3  9 12 13 14 17  7  4  5]
Incerta:  [17 20 23  5 12 13 22  0  2 10 11 14 21 18 19  1  9  3 15  4 16  8  7  6]


##### Padronização dos peridoos

Para isso será determinados alguns limites de horários para cada uma das categorias descritas acima

 - **Madrugada:** 0h às 5h
 - **Manhã:** 5h às 12h
 - **Tarde:** 12h às 18h
 - **Noite:** 18h às 24h
 - **Incerta:** Será mantido o registro original

In [91]:
df['PERIODOOCORRENCIA'] = np.where((df['DATA_HORA_OCORRENCIA'].dt.hour >= 0) & (df['DATA_HORA_OCORRENCIA'].dt.hour < 5),
                                  'MADRUGADA', df['PERIODOOCORRENCIA'])

df['PERIODOOCORRENCIA'] = np.where((df['DATA_HORA_OCORRENCIA'].dt.hour >= 5) & (df['DATA_HORA_OCORRENCIA'].dt.hour < 12),
                                  'MANHA', df['PERIODOOCORRENCIA'])

df['PERIODOOCORRENCIA'] = np.where((df['DATA_HORA_OCORRENCIA'].dt.hour >= 12) & (df['DATA_HORA_OCORRENCIA'].dt.hour < 18),
                                  'TARDE', df['PERIODOOCORRENCIA'])

df['PERIODOOCORRENCIA'] = np.where((df['DATA_HORA_OCORRENCIA'].dt.hour >= 18) & (df['DATA_HORA_OCORRENCIA'].dt.hour <= 23),
                                  'NOITE', df['PERIODOOCORRENCIA'])


In [92]:
df['PERIODOOCORRENCIA'].loc[dfinc.index] = 'EM HORA INCERTA'

c:\users\caior\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [93]:
df['PERIODOOCORRENCIA'].unique()

array(['NOITE', 'MADRUGADA', 'MANHA', 'TARDE', 'EM HORA INCERTA'],
      dtype=object)

In [94]:
df.PERIODOOCORRENCIA.value_counts()

NOITE              2441907
MANHA              1529095
TARDE              1448565
MADRUGADA           681020
EM HORA INCERTA     153224
Name: PERIODOOCORRENCIA, dtype: int64

## Salvar o arquivo em um  arquivo PARQUET

In [95]:
df.to_parquet('data_cleaned/ssp_cleaned.parquet', index=False)